source/endpoints_regex.c

```diff
@@ -0,0 +1,312 @@
+    uint8_t ch = text->ptr[0];
+    switch (symbol->type) {
+        case AWS_ENDPOINTS_REGEX_SYMBOL_ALPHA:
+            return aws_isalpha(ch);
+        case AWS_ENDPOINTS_REGEX_SYMBOL_DIGIT:
+            return aws_isdigit(ch);
+        case AWS_ENDPOINTS_REGEX_SYMBOL_CHAR:
+            return ch == symbol->info.ch;
+        case AWS_ENDPOINTS_REGEX_SYMBOL_DOT:
+            return true;
+        default:
+            AWS_FATAL_ASSERT(true);
+    }
+
+    return false;
+}
+
+static bool s_match_star(struct aws_endpoint_regex_symbol *symbol, struct aws_byte_cursor *text) {
+    while (s_match_one(symbol, text)) {
+        aws_byte_cursor_advance(text, 1);
+    }
+
+    return true;
+}
+
+static bool s_match_plus(struct aws_endpoint_regex_symbol *symbol, struct aws_byte_cursor *text) {
+    if (!s_match_one(symbol, text)) {
+        return false;
+    }
+
+    aws_byte_cursor_advance(text, 1);
+    return s_match_star(symbol, text);
+}
+
+int aws_endpoint_regex_match(struct aws_endpoint_regex *regex, struct aws_byte_cursor text) {
+    AWS_PRECONDITION(regex);
+
+    if (text.len == 0 || text.len > s_max_text_length) {
+        AWS_LOGF_ERROR(
+            AWS_LS_SDKUTILS_ENDPOINTS_REGEX, "Invalid text size. Must be between 1 and %zu", s_max_text_length);
+        return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    }
+
+    struct aws_array_list *symbols = (struct aws_array_list *)regex;
+
+    for (size_t i = 0; i < aws_array_list_length(symbols); ++i) {
+        struct aws_endpoint_regex_symbol *symbol = NULL;
+        aws_array_list_get_at_ptr(symbols, (void **)&symbol, i);
+
+        /* looks forward to check if symbol has * or + modifier */
+        if (i + 1 < aws_array_list_length(symbols)) {
+            struct aws_endpoint_regex_symbol *next_symbol = NULL;
+            aws_array_list_get_at_ptr(symbols, (void **)&next_symbol, i + 1);
+
+            if (next_symbol->type == AWS_ENDPOINTS_REGEX_SYMBOL_STAR ||
+                next_symbol->type == AWS_ENDPOINTS_REGEX_SYMBOL_PLUS) {
+                if (next_symbol->type == AWS_ENDPOINTS_REGEX_SYMBOL_STAR && !s_match_star(symbol, &text)) {
+                    return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_REGEX_NO_MATCH);
+                } else if (next_symbol->type == AWS_ENDPOINTS_REGEX_SYMBOL_PLUS && !s_match_plus(symbol, &text)) {
+                    return aws_raise_error(AWS_ERROR_SDKUTILS_ENDPOINTS_REGEX_NO_MATCH);
+                }
+                ++i;
+                continue;
+            }
+        }
+
+        switch (symbol->type) {
+            case AWS_ENDPOINTS_REGEX_SYMBOL_ALTERNATION_GROUP: {
+                struct aws_byte_cursor variant = {0};
+                struct aws_byte_cursor alternation = aws_byte_cursor_from_string(symbol->info.alternation);
+                size_t chars_in_match = 0;
+                while (aws_byte_cursor_next_split(&alternation, '|', &variant)) {
+                    if (aws_byte_cursor_starts_with(&text, &variant)) {
```

Comment:
Mention in the rules that it just matches the first alternation and will fail for something like regex = (ab|abc)d  and text = abcd

Reply:
Mentioning this in the rules in a .c file seems insufficient. If we're not going to handle certain patterns, we should detect and error out in regex_new()

But we can easily handle this. Simply scan over every variant, and use whichever one was longest.